File: solutions/ocaml/066-construct-binary-tree-from-preorder-and-inorder-traversal.ml
(*
  Problem 66: Construct Binary Tree from Preorder and Inorder Traversal (LeetCode 105)
  Difficulty: Med
  Language: OCaml
*)
let rec build_tree preorder inorder = match preorder with
  | [] -> Nil
  | r :: pre ->
    let rec split = function
      | [] -> ([], [])
      | x :: xs when x = r -> ([], xs)
      | x :: xs -> let (l, rr) = split xs in (x :: l, rr)
    in
    let (left_in, right_in) = split inorder in
    let n = List.length left_in in
    let left_pre = List.filteri (fun i _ -> i < n) pre in
    let right_pre = List.filteri (fun i _ -> i >= n) pre in
    Node (r, build_tree left_pre left_in, build_tree right_pre right_in)

File: solutions/elixir/066-construct-binary-tree-from-preorder-and-inorder-traversal.ex
# Problem 66: Construct Binary Tree from Preorder and Inorder Traversal (LeetCode 105)
# Difficulty: Med
# Language: Elixir
# 
defmodule BuildTree do
  def build([], _), do: nil
  def build([r | pre], ino) do
    {left_in, [_ | right_in]} = Enum.split_while(ino, &(&1 != r))
    n = length(left_in)
    {left_pre, right_pre} = Enum.split(pre, n)
    {r, build(left_pre, left_in), build(right_pre, right_in)}
  end
end
